### ISSUE TITLE
Incorrect error reporting with object spread operator

### ISSUE THREAD

**Issue by sliminality** [microsoft/TypeScript#60675]
### 🔎 Search Terms

`spread operator destructuring error`, `spread error message`

### 🕗 Version & Regression Information

- This changed between versions [4.2.3](https://www.typescriptlang.org/play/?ts=4.2.3#code/C4TwDgpgBAogbhAdsACgQxAGwPZoCYDOUAvFAN4BQU1UAxtgLZiYTAQDKYATgJbIBc5KjRFoArsEZpgPbIgCSeQQWC9EAcwA0wkdQJg0tCIuWq+WnVAC+2kQSR4AshAIE06iIMq6aDF249TNQsRGworCgpQSCgAYUZmVgg8eCRgAB4AFSgIAA82REIoAGsIEGwAM1gEZHQsXEIAPhIhEQBtGCg+KEyAXS9LEWjPWFsfKAN6-EFU2owcfAIO3rGfNDAwADk0PyDzVetwtr7IirFEWhk5KGwELgB3XjYAQQ3tvyyc-IciUvKq2aoeYNAiNAAUlgItAAFskxCwlHEEiw2CkahlMo1tABKQTxJgo5KArLNbw0eiIFTkKDrLY7CCaKAAOhZXBcwGsLShsLw8OSlgA9AKoAB1CA8Lh4HJcLjYLhQPyudzQWFsxkqNBcGQaLqIKAAFiZAGZLGzgGIuHqyT4WUy2SoDqI3vTBAAiNgqV0HCJWIA) and [4.3.5](https://www.typescriptlang.org/play/?ts=4.3.5#code/C4TwDgpgBAogbhAdsACgQxAGwPZoCYDOUAvFAN4BQU1UAxtgLZiYTAQDKYATgJbIBc5KjRFoArsEZpgPbIgCSeQQWC9EAcwA0wkdQJg0tCIuWq+WnVAC+2kQSR4AshAIE06iIMq6aDF249TNQsRGworCgpQSCgAYUZmVgg8eCRgAB4AFSgIAA82REIoAGsIEGwAM1gEZHQsXEIAPhIhEQBtGCg+KEyAXS9LEWjPWFsfKAN6-EFU2owcfAIO3rGfNDAwADk0PyDzVetwtr7IirFEWhk5KGwELgB3XjYAQQ3tvyyc-IciUvKq2aoeYNAiNAAUlgItAAFskxCwlHEEiw2CkahlMo1tABKQTxJgo5KArLNbw0eiIFTkKDrLY7CCaKAAOhZXBcwGsLShsLw8OSlgA9AKoAB1CA8Lh4HJcLjYLhQPyudzQWFsxkqNBcGQaLqIKAAFiZAGZLGzgGIuHqyT4WUy2SoDqI3vTBAAiNgqV0HCJWIA)

### ⏯ Playground Link

[Playground link, 5.7.2](https://www.typescriptlang.org/play/?ts=5.7.2#code/C4TwDgpgBAogbhAdsACgQxAGwPZoCYDOUAvFAN4BQU1UAxtgLZiYTAQDKYATgJbIBc5KjRFoArsEZpgPbIgCSeQQWC9EAcwA0wkdQJg0tCIuWq+WnVAC+2kQSR4AshAIE06iIMq6aDF249TNQsRGworCgpQSCgAYUZmVgg8eCRgAB4AFSgIAA82REIoAGsIEGwAM1gEZHQsXEIAPhIhEQBtGCg+KEyAXS9LEWjPWFsfKAN6-EFU2owcfAIO3rGfNDAwADk0PyDzVetwtr7IirFEWhk5KGwELgB3XjYAQQ3tvyyc-IciUvKq2aoeYNAiNAAUlgItAAFskxCwlHEEiw2CkahlMo1tABKQTxJgo5KArLNbw0eiIFTkKDrLY7CCaKAAOhZXBcwGsLShsLw8OSlgA9AKoAB1CA8Lh4HJcLjYLhQPyudzQWFsxkqNBcGQaLqIKAAFiZAGZLGzgGIuHqyT4WUy2SoDqI3vTBAAiNgqV0HCJWIA)

### 💻 Code

```ts
type EventPayloads = {
    completeSprint: {
        automationId: string,
        spaceId: string,
    },
    sendMessage: {
        message: string,
    },
}

type CompletedEvent<T extends keyof EventPayloads> = {
    [E in T]: {
        type: E,
        payload: EventPayloads[E],
        appName: string,
    }
}[T]

function overwriteAppName<T extends keyof EventPayloads>(
    scheduled: CompletedEvent<T>,
): CompletedEvent<T> {
    const { appName, ...rest } = scheduled
    // Weird error message here, starting in 4.3
    return {
        ...rest,
        appName: "test",
    }
}
```

### 🙁 Actual behavior

The `return` statement has the following error:
```
Type 'Omit<CompletedEvent<T>, "appName"> & { appName: string; }' is not assignable to type 'CompletedEvent<T>'.
  Types of property 'payload' are incompatible.
    Type 'CompletedEvent<T>["payload"]' is not assignable to type 'EventPayloads[T]'.
      Type '{ type: T; payload: EventPayloads[T]; appName: string; }' is missing the following properties from type 'EventPayloads': completeSprint, sendMessage(2322)
```
The first three lines make sense, but the last line is bizarre enough that I suspect it's a bug in either the reporting/checking logic.

Given that `EventPayloads` and `CompletedEvent` are completely different objects, this error made me think I had missed an indexing step somewhere, but after rereading the code multiple times I'm pretty sure I didn't (it seems like a type error in the type error).

Also, my best possible explanation for the reduction from
```
    Type 'CompletedEvent<T>["payload"]' is not assignable to type 'EventPayloads[T]'.
```
to
```
      Type '{ type: T; payload: EventPayloads[T]; appName: string; }' is missing the following properties from type 'EventPayloads': completeSprint, sendMessage(2322)
```
is that the algorithm eliminated the indexed access on both sides, so that

| | `CompletedEvent<T>["payload"]` | `EventPayloads[T]` |
|-|-|-|
|becomes | `CompletedEvent<T>` | `EventPayloads` |

but this is clearly incorrect considering `T != "payload"`.

### 🙂 Expected behavior

Given the penultimate line of the error, I would expect the final line to compare payload types, i.e.
```ts
{ automationId: string, spaceId: string }
```
vs. 
```ts
{ message: string }
```

For what it's worth, this is the error from [4.2.3](https://www.typescriptlang.org/play/?ts=4.2.3#code/C4TwDgpgBAogbhAdsACgQxAGwPZoCYDOUAvFAN4BQU1UAxtgLZiYTAQDKYATgJbIBc5KjRFoArsEZpgPbIgCSeQQWC9EAcwA0wkdQJg0tCIuWq+WnVAC+2kQSR4AshAIE06iIMq6aDF249TNQsRGworCgpQSCgAYUZmVgg8eCRgAB4AFSgIAA82REIoAGsIEGwAM1gEZHQsXEIAPhIhEQBtGCg+KEyAXS9LEWjPWFsfKAN6-EFU2owcfAIO3rGfNDAwADk0PyDzVetwtr7IirFEWhk5KGwELgB3XjYAQQ3tvyyc-IciUvKq2aoeYNAiNAAUlgItAAFskxCwlHEEiw2CkahlMo1tABKQTxJgo5KArLNbw0eiIFTkKDrLY7CCaKAAOhZXBcwGsLShsLw8OSlgA9AKoAB1CA8Lh4HJcLjYLhQPyudzQWFsxkqNBcGQaLqIKAAFiZAGZLGzgGIuHqyT4WUy2SoDqI3vTBAAiNgqV0HCJWIA) and earlier:
```
Type 'Omit<CompletedEvent<T>, "appName"> & { appName: string; }' is not assignable to type 'CompletedEvent<T>'.
  Types of property 'payload' are incompatible.
    Type 'CompletedEvent<T>["payload"]' is not assignable to type 'EventPayloads[T]'.
      Type '{ type: T; payload: EventPayloads[T]; appName: string; }' is missing the following properties from type 'EventPayloads': completeSprint, sendMessage
        Type 'CompletedEvent<T>["payload"]' is not assignable to type '{ automationId: string; spaceId: string; } & { message: string; }'.
          Type 'EventPayloads[T]' is not assignable to type '{ automationId: string; spaceId: string; } & { message: string; }'.
            Type '{ automationId: string; spaceId: string; } | { message: string; }' is not assignable to type '{ automationId: string; spaceId: string; } & { message: string; }'.
              Type '{ automationId: string; spaceId: string; }' is not assignable to type '{ automationId: string; spaceId: string; } & { message: string; }'.
                Property 'message' is missing in type '{ automationId: string; spaceId: string; }' but required in type '{ message: string; }'.
                  Type 'EventPayloads[T]' is not assignable to type '{ automationId: string; spaceId: string; }'.
                    Type 'CompletedEvent<T>["payload"]' is not assignable to type '{ automationId: string; spaceId: string; }'.
                      Type 'EventPayloads[T]' is not assignable to type '{ automationId: string; spaceId: string; }'.
                        Type '{ automationId: string; spaceId: string; } | { message: string; }' is not assignable to type '{ automationId: string; spaceId: string; }'.
                          Type '{ message: string; }' is missing the following properties from type '{ automationId: string; spaceId: string; }': automationId, spaceId(2322)
```

### Additional information about the issue

I've tagged the issue with `object spread operator` because [removing the spreads resolves the error](https://www.typescriptlang.org/play/?ts=5.7.2#code/C4TwDgpgBAogbhAdsACgQxAGwPZoCYDOUAvFAN4BQU1UAxtgLZiYTAQDKYATgJbIBc5KjRFoArsEZpgPbIgCSeQQWC9EAcwA0wkdQJg0tCIuWq+WnVAC+2kQSR4AshAIE06iIMq6aDF249TNQsRGworCgpQSCgAYUZmVgg8eCRgAB4AFSgIAA82REIoAGsIEGwAM1gEZHQsXEIAPhIhEQBtGCg+KEyAXS9LEWjPWFsfKAN6-EFU2owcfAIO3rGfNDAwADk0PyDzVetwtr7IirFEWhk5KGwELgB3XjYAQQ3tvyyc-IciUvKq2aoeYNAiNAAUlgItAAFskxCwlHEEiw2CkahlMo1tABKQTxJgo5KArLNbw0eiIFTkKDDTQTYH4awtKGwvDw5KWAD0nKgAHUeMBodgJFB9FwIPg6YLoBAuFxsFwoOpsC4oGh7hgAHSWcXAMRcRCtca0wY0SYLPAHURvHYjABEbBUdoOESsQA). But I don't actually know whether the bug is intrinsic to spreads.

**Comment by jfet97** [comment]
I played a little bit with it to get what's happening under the hood, testing the following code:

```ts
type EventPayloads = {
    completeSprint: {
        automationId: string,
        spaceId: string,
    },
    sendMessage: {
        message: string,
    },
}

type CompletedEvent<T extends keyof EventPayloads> = {
    [E in T]: {
        type: E,
        payload: EventPayloads[E],
        appName: string,
    }
}[T]

function overwriteAppName<U extends keyof EventPayloads>(
    scheduled: CompletedEvent<U>,
) {
    const { appName, ...rest } = scheduled

    // scheduled.payload has type EventPayloads[U]
    // rest.payload has type CompletedEvent<U>["payload"]
    scheduled.payload = rest.payload // fails :(
    rest.payload = scheduled.payload // passes :)
}
```

Where the first assignment fails with the same final error message as yours, while the second one is fine for TS.

Internally the checker passes trough a point containing this comment:
```ts
// Relate components directly before falling back to constraint relationships
// A type S[K] is related to a type T[J] if S is related to T and K is related to J.
```
As you noticed, it strips away the indexes to check if `CompletedEvent<U>` is related with `EventPayloads`, but that's not the case of course. I would dare to say that the main reason behind this is that the `scheduled.payload` is eagerly resolved into `EventPayloads[U]` while `rest.payload` is not, so when it checks that assignment it ends up checking the relation between something that has already been computed and another thing that's still deferred. You do not directly access to `rest.payload`, but it seems TS does something similar when checking what you are returning vs the expected return type.

This part fails for both assignments.

After that there is a constraint relationships checking:
```ts
// A type S is related to a type T[K] if S is related to C, where C is the base
// constraint of T[K] for writing.
```
In few words, it computes the constraint of the __target__ and tries again to assign the __source__ to it.\
When it comes to the second assignment, the constraint of `rest.payload` is computed and the type of `scheduled.payload` is checked against it. The constraint of `rest.payload` is taken from the constraint of both its `indexType` and its `objectType`. It ends up computing  something like `{ type: U; payload: EventPayloads[U]; appName: string; }["payload"]` that is `EventPalyloads[U]`. So everything is fine now, but only for the second assignment.\
For the first one this part is useless because the constraint of `scheduled.payload` is `{ automationId: string; spaceId: string; } & { message: string; }`.

**Comment by jfet97** [comment]
@Andarist Yeah that's the missing reduction step on `rest.payload`. Actually I'm wondering if that should be attempted here as a fallback as you are doing or if it would be better to have it already computed beforehand.

**Comment by Andarist** [comment]
> Actually I'm wondering if that should be attempted here as a fallback as you are doing or if it would be better/possible to have it already computed beforehand, so that the type assigned to rest.payload is just EventPayloads[U].

I'm pretty sure it's deferred on purpose and that it likely helps some other relationships to be checked. Once you drill into such indexed access you might end up with a union or smth and it's computationally more expensive to prove that such union is assignable to certain targets. It's the reason why the indexed mapped type accesses (the pattern you like ;p) can be type-checked while the equivalent unions can't be.

**Comment by jfet97** [comment]
> the pattern you like ;p

I may have dug into this just because of how OP defined `CompletedEvent`, yeah 👀
